### PR TITLE
Expose disable dynamic stats

### PR DESCRIPTION
**Pull request by ashishb-solo**
# Description

Expose an option to disable dynamic stats for benchmarking/profiling purposes.

## API changes
- Added `dynamic_stats` field to the router message

## Docs changes
- Inline protobuf documentation for dynamic stats field

# Context

This request came from a user who was interested in understanding the differences in performance between http1 and http2.

## Testing steps

Not yet implemented

## Notes for reviewers

# Checklist:

- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: projects/gloo/api/v1/options/router/router.proto [hunk 1]
```diff
@@ -9,4 +9,7 @@ import "google/protobuf/wrappers.proto";
 message Router{ 
     // Do not add any additional x-envoy- headers to requests or responses. This only affects the router filter generated x-envoy- headers, other Envoy filters and the HTTP connection manager may continue to set x-envoy- headers.
     google.protobuf.BoolValue suppress_envoy_headers = 1;
-}
\ No newline at end of file
+
+    // Disable dynamic stats. Defaults to false. This should only be set to true for profiling or benchmarking scenarios - it is strongly recommended to NOT set this in production
```
Comment (ben-taussig-solo) [inline projects/gloo/api/v1/options/router/router.proto:13]: - I think this should say `enable` dynamic stats, but I am not sure
 - Can we update this to say something like "Set to true to enable dynamic stats" -- I think at the moment it is slightly unclear what will happen if the value is set to true/false

File: projects/gloo/api/v1/options/router/router.proto [hunk 2]
```diff
@@ -9,4 +9,7 @@ import "google/protobuf/wrappers.proto";
...
+    google.protobuf.BoolValue dynamic_stats = 2;
```
Comment (ben-taussig-solo) [inline projects/gloo/api/v1/options/router/router.proto:14]: given that this defaults to false and that we don't seem to be making use of null values, can we just use a bool here instead of a `google.protobuf.BoolValue`?

File: projects/gloo/pkg/translator/listener_subsystem_test.go [hunk 3]
```diff
@@ -176,13 +177,16 @@ var _ = Describe("Listener Subsystem", func() {
 			},
 		),
 		Entry(
-			"Add suppress envoy headers to the router",
+			"Set dynamic_stats to false and suppress_envoy_headers to true",
```
Comment (ben-taussig-solo) [inline projects/gloo/pkg/translator/listener_subsystem_test.go:180]: can we decouple the preexisting test around supress_envoy_headers from the new testing around dynamic stats? correct me if I'm wrong but they don't seem to affect each other

File: projects/gloo/pkg/translator/network_filters.go [hunk 4]
```diff
@@ -240,6 +240,8 @@ func (h *hcmNetworkFilterTranslator) computeHttpFilters(params plugins.Params) [
 		routerV3.SuppressEnvoyHeaders = true
 	}
 
+	routerV3.DynamicStats = h.listener.GetOptions().GetRouter().GetDynamicStats()
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/network_filters.go:243]: nit: I see that we are following the existing convention of defining the router options in this network_filters file. In general, for any filter that has any amount of complexity, we've tended to define a plugin that holds all the logic. Do you think since we're now adding more fields, it's worth it to introduce a router plugin to hold this logic?

File: projects/gloo/api/v1/options/router/router.proto (hunk 1 above)
Comment (sam-heilbron) [reply projects/gloo/api/v1/options/router/router.proto:13]: I agree with Ben. I find the name of the field and the comment "disable dynamic stats" to be a bit confusing

File: projects/gloo/pkg/translator/listener_subsystem_test.go [hunk 5]
```diff
@@ -202,7 +206,48 @@ var _ = Describe("Listener Subsystem", func() {
 				}},
 			},
 			func(listener *envoy_config_listener_v3.Listener, routeConfigs []*envoy_config_route_v3.RouteConfiguration) {
-				By("Should be able to add and translate the router to an envoy config")
+				By("configuring the envoy router from gloo settings")
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/listener_subsystem_test.go:209]: nit: I know we're following prior art, so feel free to disregard.

It feels like the listener_subsystem_test isn't the appropriate place for these tests. Per the comment:
```
// These tests validate that the ListenerSubsystemTranslatorFactory produces Translators
// which in turn create Envoy Listeners and RouteConfigurations with expected values
// The tests are non-exhaustive, as we expect each translator to more rigorously test the
// edge cases. Instead, these tests focus on the high level Envoy resources that are created.
```

It seems like these tests either belong in a network_filters_test which proves that router filter is added, or better yet, a plugin that performs the translation and corresponding unit tests to prove it out

File: projects/gloo/pkg/translator/listener_subsystem_test.go [hunk 6]
```diff
@@ -202,7 +206,48 @@ var _ = Describe("Listener Subsystem", func() {
...
+				hcm := &envoy_http_connection_manager_v3.HttpConnectionManager{}
+				err = translator.ParseTypedConfig(hcmFilter, hcm)
+				Expect(err).NotTo(HaveOccurred())
+				Expect(hcm.HttpFilters).To(HaveLen(2))
+
+				routeFilter := hcm.GetHttpFilters()[1]
+				typedRouterFilter := routerv3.Router{}
+				routeFilter.GetTypedConfig().UnmarshalTo(&typedRouterFilter)
+				Expect(typedRouterFilter.GetDynamicStats().GetValue()).To(BeFalse())
+				Expect(typedRouterFilter.SuppressEnvoyHeaders).To(BeTrue())
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/listener_subsystem_test.go:224]: Let's the use the Getter method for SuppressEnvoyHeaders

File: projects/gloo/pkg/translator/listener_subsystem_test.go [hunk 7]
```diff
@@ -202,7 +206,48 @@ var _ = Describe("Listener Subsystem", func() {
...
+			},
+		),
+		Entry(
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/listener_subsystem_test.go:227]: nit: It feels like we could simplify these tests into a simple Table driven test that has cases for:
- unset values
- set true
- set false

File: projects/gloo/api/v1/options/router/router.proto (hunk 1 above)
Comment (ashishb-solo) [reply projects/gloo/api/v1/options/router/router.proto]: Thanks, I was struggling on how to make it clear. I took another stab at it in 78e943e6fddc54a601c20dd0de3c4227667d092d - could you let me know if you think it's an improvement? Hopefully it's more explicit about what the possible states do (true, false, nil)

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 7 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go:227]: Sounds like a good idea to me. Were you thinking of doing this in a separate table, or is it possible to have nested table tests in ginkgo/gomega?

File: projects/gloo/pkg/translator/network_filters.go (hunk 4 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/network_filters.go:243]: I like the idea and thought the same thing while writing this, but think it might be a better idea to do this some other time due to (a) time constraints and (b) the fact that the complexity of the current solution is still fairly manageable (in my opinion)

I think I'll just add a TODO in this translation area for the time being, since there are more complicated options in the router filter that we may want to implement at some point. Do you think that's fair for now?

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 7 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go]: Created a new table test here: https://github.com/solo-io/gloo/pull/9260/files#diff-1f2deb4318fe8a371631759130cc64f355caf466ded0bd46a996d5f71df48e85R65

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 6 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go]: This moved around a bit in the test refactor, but you can find the updated version here: https://github.com/solo-io/gloo/pull/9260/files#diff-1f2deb4318fe8a371631759130cc64f355caf466ded0bd46a996d5f71df48e85R151

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 5 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go]: It took quite a few commits to get this in, but I've moved these tests out to here: https://github.com/solo-io/gloo/pull/9260/files#diff-1f2deb4318fe8a371631759130cc64f355caf466ded0bd46a996d5f71df48e85R1

File: projects/gloo/pkg/translator/network_filters.go (hunk 4 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/network_filters.go:245]: i put in a TODO here, let me know how you feel: d9283ab417f7803b4d0396765cc803d6889087c8

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 3 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go]: broken out tests can be found here: https://github.com/solo-io/gloo/pull/9260/files#diff-1f2deb4318fe8a371631759130cc64f355caf466ded0bd46a996d5f71df48e85R143

File: projects/gloo/api/v1/options/router/router.proto (hunk 1 above)
Comment (ben-taussig-solo) [reply projects/gloo/api/v1/options/router/router.proto]: The update looks much clearer, thanks!

File: projects/gloo/pkg/translator/listener_subsystem_test.go (hunk 3 above)
Comment (ben-taussig-solo) [reply projects/gloo/pkg/translator/listener_subsystem_test.go]: those new table entries are very easy to read, thanks for this!

File: projects/gloo/api/v1/options/router/router.proto (hunk 2 above)
Comment (ashishb-solo) [reply projects/gloo/api/v1/options/router/router.proto:18]: I agree that it's a bit silly to have two values that do the same thing. I'm inclined to just leave it as is and stay as close as possible to the Envoy API, but happy to change it if we see a good reason to do so

File: projects/gloo/pkg/translator/network_filters_test.go [hunk 8]
```diff
@@ -0,0 +1,174 @@
...
+		}
+
+		translatorFactory = translator.NewListenerSubsystemTranslatorFactory(pluginRegistry, sslutils.NewSslConfigTranslator())
+	})
+
+	AfterEach(func() {
+		cancel()
+	})
+
+	DescribeTable("GetAggregateListenerTranslators (success)",
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/network_filters_test.go:65]: nit: Given that these tests are in a new file (network_filters) and are specifically testing the behavior of the httpNetworkFilterTranslator, should we change this name, and perhaps change the inputs to call `ComputeNetworkFilters` ?

File: projects/gloo/pkg/translator/network_filters_test.go [hunk 9]
```diff
@@ -0,0 +1,174 @@
...
+			hcm := &envoy_http_connection_manager_v3.HttpConnectionManager{}
+			err = translator.ParseTypedConfig(hcmFilter, hcm)
+			Expect(err).NotTo(HaveOccurred())
+			Expect(hcm.HttpFilters).To(HaveLen(1))
+
+			routeFilter := hcm.GetHttpFilters()[0]
+			typedRouterFilter := routerv3.Router{}
+			routeFilter.GetTypedConfig().UnmarshalTo(&typedRouterFilter)
+			// Perform assertions on generated Envoy router filter
+			assertionHandler(&typedRouterFilter)
```
Comment (sam-heilbron) [inline projects/gloo/pkg/translator/network_filters_test.go:140]: nit: The solution you have works well. An alternative, is to just pass a GomegaMatcher, and then call:
```
Expect(typedRouterFilter).To(expectedMatcher)
```

An example of that is here: https://github.com/solo-io/gloo/blob/main/projects/gloo/pkg/syncer/setup/extensions_test.go#L17

Not a requirement, just sharing an alternative way

**Review by sam-heilbron — APPROVED** [review]

File: projects/gloo/pkg/translator/network_filters_test.go (hunk 8 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/network_filters_test.go]: Nice catch on the table name. I've fixed that.

Regarding calling `ComputeNetworkFilters`, I would need an instance of `hcmNetworkFilterTranslator` and I don't see any code that directly creates or instantiates that. So I'm inclined to leave it as is given the amount of work involved in doing that (we still have to push a release out based on this, so I'd ideally like to merge this today/early tomorrow).

However, I did simplify the test down a bit by switching from using an AggregateListener to a HttpListener and this had a modest reduction on the size of the table function.

Both of these changes can be found in https://github.com/solo-io/gloo/pull/9260/commits/0ff31e7f157b9422e82b1a52e155025bb5b2ba5e

File: projects/gloo/pkg/translator/network_filters_test.go (hunk 9 above)
Comment (ashishb-solo) [reply projects/gloo/pkg/translator/network_filters_test.go:128]: Thanks for the heads up. I like the way it is because it allows us to reuse the same architecture to test different fields of the `typedRouterFilter` struct (suppress headers and dynamic stats). I wouldn't be surprised if Gomega provides a way to specify which field the matcher ought to check on, but I would prefer to leave it as is now

File: projects/gloo/pkg/translator/network_filters_test.go (hunk 8 above)
Comment (sam-heilbron) [reply projects/gloo/pkg/translator/network_filters_test.go]: Yeah I think for constructing the hcmNetworkFilterTranstlator, you could use `NewHttpListenerNetworkFilterTranslator`. Not necessary, I think what you have makes sense

File: projects/gloo/pkg/translator/network_filters_test.go (hunk 9 above)
Comment (sam-heilbron) [reply projects/gloo/pkg/translator/network_filters_test.go:128]: Yeah that's fair. I think the way you could solve this, using a GomegaMatcher, would be something like:
```
Expect(typedRouterFilter).To(matchers.MatchProto(&routerV1.Router{
	SuppressEnvoyHeaders: &wrappers.BoolValue{
		Value: true,
	},
}))
```

Where MatchProto comes from: https://github.com/solo-io/solo-kit/blob/main/test/matchers/proto.go#L40. And allows us to compare top level fields of a proto message.

**Review by ben-taussig-solo — APPROVED** [review]
